daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala

```diff
@@ -20,42 +20,45 @@ private[speedy] object ClosureConversion {
 
   private[speedy] def closureConvert(source0: source.SExpr): target.SExpr = {
 
-    // TODO: Recode the 'Env' management to avoid the polynomial-complexity of 'shift'. Issue #11830
-    case class Env(depth: Int, mapping: Map[Int, target.SELoc]) {
+    case class Abs(a: Int) // absolute variable index, determined by tracking sourceDepth
 
-      def lookup(i: Int): target.SELoc =
-        mapping.get(i) match {
+    case class Env(sourceDepth: Int, mapping: Map[Abs, target.SELoc], targetDepth: Int) {
+
+      def lookup(abs: Abs): target.SELoc = {
+        mapping.get(abs) match {
           case Some(loc) => loc
           case None =>
-            throw sys.error(s"lookup($i),in:$mapping")
+            throw sys.error(s"lookup($abs),in:$mapping")
         }
+      }
 
-      def shift(n: Int): Env = {
-        // We just update the keys of the map (the relative-indexes from the original SEVar)
-        val m1 = mapping.map { case (k, loc) => (n + k, loc) }
-        // And create mappings for the `n` new stack items
-        val m2 = (1 to n).view.map { rel =>
-          val abs = this.depth + n - rel
-          (rel, target.SELocAbsoluteS(abs))
+      def extend(n: Int): Env = {
+        // Create mappings for `n` new stack items, and combine with the (unshifted!) existing mapping.
+        val m2 = (1 to n).view.map { i =>
+          val abs = Abs(this.sourceDepth + (n - i))
+          (abs, target.SELocAbsoluteS(this.targetDepth + n - i))
```

Comment:
curious to know why we subtract `i` from `n`, and not use something like :
```suggestion
        val m2 = (0 until n).view.map { i =>
          val abs = Abs(this.sourceDepth + i)
          (abs, target.SELocAbsoluteS(this.targetDepth + i))
```

Reply:
Yeah. This was a hang-over from the old _relative_ indexing which started from `1`.
The new _absolute_ indexing starts from `0`

I simplified the code along the lines you suggest in both the `extend` and `absBody` methods.
Much simpler now. Thanks!